Mathlib/Topology/Algebra/Order/LiminfLimsup.lean

```diff
@@ -520,6 +520,121 @@ theorem limsup_eq_tendsto_sum_indicator_atTop (R : Type*) [StrictOrderedSemiring
 
 end Indicator
 
+section LiminfLimsupAdd
+
+variable [AddCommGroup α] [ConditionallyCompleteLinearOrder α] [DenselyOrdered α]
+  [CovariantClass α α (fun a b ↦ a + b) fun x1 x2 ↦ x1 ≤ x2]
+  {f : Filter ι} [f.NeBot] {u v : ι → α}
+
+lemma le_limsup_add (h₁ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f u := by isBoundedDefault)
+    (h₂ : IsCoboundedUnder (fun x1 x2 ↦ x1 ≤ x2) f u := by isBoundedDefault)
+    (h₃ : IsBoundedUnder (fun x1 x2 ↦ x1 ≤ x2) f v := by isBoundedDefault)
+    (h₄ : IsBoundedUnder (fun x1 x2 ↦ x1 ≥ x2) f v := by isBoundedDefault) :
+    (limsup u f) + liminf v f ≤ limsup (u + v) f := by
+  have h := isCoboundedUnder_le_add h₄ h₂ -- These `have` tactic improve performance.
+  have h' := isBoundedUnder_le_add h₃ h₁
+  rw [add_comm] at h h'
+  exact add_le_of_forall_lt fun a a_u b b_v ↦
+    (le_limsup_iff h h').2 fun c c_ab ↦ ((frequently_lt_of_lt_limsup h₂ a_u).and_eventually
+    (eventually_lt_of_lt_liminf b_v h₄)).mono  fun _ ab_x ↦ c_ab.trans (add_lt_add ab_x.1 ab_x.2)
```

Comment:
Do you think you could rewrite this part of the proof using `filter_upwards`, for readability?

Reply:
I think the main problem is that `filter_upwards` works with eventual goals, and we would like to have it work with frequent goals when you have at most one frequent hypothesis (something I discussed yesterday with Patrick Massot).

In the meantime, it will just be tinkering; I don't see a nice way to use `filter_upwards` (we need to isolate the eventual part of the proof, but then `filter_upwards` already applies `.mono`... Doable, but not sure it would be more readable).

As a compromise, I broke up the proof term using `refine`s, which makes it easier to understand in tactical state, and easier to modify if such an improvement of `filter_upwards` is done.